_posts/2018-01-24-kubernetes-and-hokusai.markdown

```diff
@@ -0,0 +1,134 @@
+
+Next up was to use Hokusai in our CI environment. [Circle CI 2.0][circle] is very Docker-oriented, so we set up everything using their [Workflows][]. This is a much higher level of abstraction for CI configuration than I'm used to, but I got the hang of it quickly. I created a job to run RSpec tests through Hokusai, a job to run [Danger][], a job to build and push a Docker image to our S3 bucket, and a job to deploy that image to the Kubernetes cluster. Finally, I added the workflows to build and deploy automatically after successful builds on the `master` branch.
+
+Here's a slightly redacted copy of our Circle config:
+
+```yaml
+version: 2
+jobs:
+  test:
+    docker:
+      - image: artsy/hokusai:0.4.0
+    working_directory: ~/REPO_NAME
+    steps:
+      - add_ssh_keys
+      - checkout
+      - setup_remote_docker
+      - run:
+          name: Test
+          command: hokusai test
+  danger:
+    docker:
+      - image: circleci/ruby:2.5.0
+    working_directory: ~/apogee
+    steps:
+      - checkout
+      - restore_cache:
+          keys:
+          - v1-dependencies-{{ checksum "Gemfile.lock" }}
+          - v1-dependencies-
+      - run:
+          name: Install Dependencies
+          command: bundle install --with=ci --without development test --path vendor/bundle
+      - save_cache:
+          paths:
+            - ./vendor/bundle
+          key: v1-dependencies-{{ checksum "Gemfile.lock" }}
+      - run:
+          name: Danger
+          command: bundle exec danger
+  push:
+    docker:
+      - image: artsy/hokusai:0.4.0
+    steps:
+      - add_ssh_keys
+      - checkout
+      - setup_remote_docker
+      - run:
+          name: Push
+          command: hokusai registry push --tag $CIRCLE_SHA1 --force --overwrite
+  deploy:
+    docker:
+      - image: artsy/hokusai:0.4.0
+    steps:
+      - add_ssh_keys
+      - checkout
+      - run:
+          name: Configure
+          command: hokusai configure --kubectl-version 1.6.3 --s3-bucket BUCKET_NAME --s3-key k8s/config --platform linux
+      - run:
+          name: Deploy
+          command: hokusai staging deploy $CIRCLE_SHA1
+workflows:
+  version: 2
+  default:
+    jobs:
+      - test
+      - danger:
+          filters:
+            branches:
+              ignore: master
+      - push:
+          filters:
+            branches:
+              only: master
+          requires:
+            - test
+      - deploy:
+          filters:
+            branches:
+              only: master
+          requires:
+            - push
+```
+
+The initial build on `master` built and pushed the server image, but the deploy failed. It's a Catch-22: we can't deploy until we have an image, but we only want to build images on CI, so we expect the first deploy to fail.
```

Comment:
This is something we want to address, so feedback here would be helpful.  Maybe acknowledge Hokusai is still < v1.0 at this point?

Reply:
Sure! I didn't find this to be a hindrance at all, really, but partially because I knew to expect it. I think probably the most straightforward solution might just be a helpful error message? Maybe after the failed deploy, Hokusai could check if the environment exists and provide a helpful explanation before failing? Or maybe it only prints the error message in a CI environment? In any case, I see this as more of a quirk than a bug. 

I didn't see an issue for it on Hokusai's repo, is there a public discussion you'd like linked to?

Reply:
Thanks - yes all part of the process.  I don't know the right solution tbh.  Tomorrow I'll make an issue for you to link to.  And maybe an apologetic line in that this is under development is appropriate is all.  I mean an apology [in the Platonic sense](https://en.wikipedia.org/wiki/Apology_(Plato)), as part of a discourse.

Reply:
https://github.com/artsy/hokusai/issues/50